docs/proposal/20220712-garbage-collection.md

```diff
@@ -0,0 +1,338 @@
+After it has been created
+So that i can investigate/overcome issues
+
+## Requirements
+
+### Functional
+
+<a name="FR1">FR1.</a> CAPA MUST support cleaning up of AWS resources created by the CCM for a tenant cluster when not using GitOps.
+
+<a name="FR2">FR2.</a> CAPA MUST support cleaning up of AWS resources created by the CCM for a tenant cluster when using GitOps.
+
+<a name="FR3">FR3.</a> CAPA MUST support cleaning up of AWS resources for unmanaged and managed clusters
+
+<a name="FR4">FR4.</a> CAPA MUST support a way to opt-out of garbage collection at the point of cluster creation.
+
+<a name="FR5">FR5.</a> CAPI MUST not allow me to delete a cluster fullt until garbage collection has occurred.
+
+<a name="FR6">FR6.</a> CAPA SHOULD provide a way for me to opt-in or opt-out a cluster from garbage collection AFTER it has been created.
+
+### Non-Functional
+
+<a name="NFR8">NFR8.</a> CAPA MUST be able to easily add additional AWS resource cleanup in the future.
+
+<a name="NFR9">NFR9.</a> Unit tests MUST exist for new garbage collection code.
+
+<a name="NFR10">NFR10.</a> e2e tests MUST exist for the new garbage collection code for both unmanaged and managed clusters..
+
+### Implementation Details/Notes/Constraints
+
+#### Proposed Changes
+
+In the initial implementation of garbage collection if the feature is enabled all clusters will by default be marked for garbage collection. But we will supply a means to opt-out at time of creation as per [FR4](#FR4).
+
+The proposed changes will be documented depending on if the cluster is being created/updated or deleted.
+
+> NOTE: garbage collection will be experimental initial and will be enabled via a feature flag.
+
+##### Cluster Creation
+
+The following sequence diagram depicts what will happen when you create a new cluster with CAPI/CAPA with this change. The numbers will be referenced in the following descriptions.
+
+![gc creation](20220712-garbage-collection-create.svg)
+
+###### API Changes
+
+If the garbage collection feature has been enabled via the feature flag then the point that a user can mark a cluster as opting out of garbage collection ([FR4](#FR4)) is when they apply the yaml (1). This will be accomplished by annotating the **AWSCluster** or **AWSManagedControlPlane** with the `aws.cluster.x-k8s.io/external-resource-gc` annotation and setting its value to **false**.
+
+If the `aws.cluster.x-k8s.io/external-resource-gc` annotation is absent or its value is set to **true** then the CAPA created cluster will be garbage collected.
+
+This annotation name will be in a publicly exported package.
+
+###### Controller Changes
+
+The controllers for `AWSCluster` and `AWSManagedControlPlane` will be modified so that on creation of the controllers you can indicate that the garbage collection feature flag is enabled. In [main.go](../../main.go) we will look to see if the feature flag is enabled and pass this in when creating the instances of the 2 controllers.
```

Comment:
```suggestion
The controllers for `AWSCluster` and `AWSManagedControlPlane` will be modified so that on the creation of the controllers you can indicate that the garbage collection feature flag is enabled. In [main.go](../../main.go) we will look to see if the feature flag is enabled and pass this in when creating the instances of the 2 controllers.
```

Reply:
Changed.